homeassistant/components/mqtt/valve.py

```diff
@@ -0,0 +1,386 @@
+"""Support for MQTT valve devices."""
+from __future__ import annotations
+
+from contextlib import suppress
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components import valve
+from homeassistant.components.valve import (
+    DEVICE_CLASSES_SCHEMA,
+    ValveEntity,
+    ValveEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_DEVICE_CLASS,
+    CONF_NAME,
+    CONF_OPTIMISTIC,
+    CONF_VALUE_TEMPLATE,
+    STATE_CLOSED,
+    STATE_CLOSING,
+    STATE_OPEN,
+    STATE_OPENING,
+)
+from homeassistant.core import HomeAssistant, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.util.json import JSON_DECODE_EXCEPTIONS, json_loads
+from homeassistant.util.percentage import (
+    percentage_to_ranged_value,
+    ranged_value_to_percentage,
+)
+
+from . import subscription
+from .config import MQTT_BASE_SCHEMA
+from .const import (
+    CONF_COMMAND_TEMPLATE,
+    CONF_COMMAND_TOPIC,
+    CONF_ENCODING,
+    CONF_PAYLOAD_CLOSE,
+    CONF_PAYLOAD_OPEN,
+    CONF_PAYLOAD_STOP,
+    CONF_POSITION_CLOSED,
+    CONF_POSITION_OPEN,
+    CONF_QOS,
+    CONF_RETAIN,
+    CONF_STATE_CLOSED,
+    CONF_STATE_CLOSING,
+    CONF_STATE_OPEN,
+    CONF_STATE_OPENING,
+    CONF_STATE_TOPIC,
+    DEFAULT_OPTIMISTIC,
+    DEFAULT_PAYLOAD_CLOSE,
+    DEFAULT_PAYLOAD_OPEN,
+    DEFAULT_POSITION_CLOSED,
+    DEFAULT_POSITION_OPEN,
+    DEFAULT_RETAIN,
+)
+from .debug_info import log_messages
+from .mixins import (
+    MQTT_ENTITY_COMMON_SCHEMA,
+    MqttEntity,
+    async_setup_entity_entry_helper,
+    write_state_on_attr_change,
+)
+from .models import MqttCommandTemplate, MqttValueTemplate, ReceiveMessage
+from .util import valid_publish_topic, valid_subscribe_topic
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_REPORTS_POSITION = "reports_position"
+
+DEFAULT_NAME = "MQTT Valve"
+
+MQTT_VALVE_ATTRIBUTES_BLOCKED = frozenset(
+    {
+        valve.ATTR_CURRENT_POSITION,
+    }
+)
+
+
+def _validate_valve_payload_options(config: ConfigType) -> ConfigType:
+    """Validate the use of payload close and open options."""
+    if config[CONF_REPORTS_POSITION] and (
+        CONF_PAYLOAD_CLOSE in config
+        or CONF_PAYLOAD_OPEN in config
+        or CONF_STATE_CLOSED in config
+        or CONF_STATE_OPEN in config
+    ):
```

Comment:
Maybe something like this is cleaner:
```suggestion
NO_POSITION = (CONF_PAYLOAD_CLOSE, CONF_PAYLOAD_OPEN, CONF_STATE_CLOSED, CONF_STATE_OPEN)
...
    if config[CONF_REPORTS_POSITION] and (any key in config for key in NO_POSITION):
```